rootston/input_method.c

```diff
@@ -0,0 +1,65 @@
+#define _POSIX_C_SOURCE 199309L
+
+#include <signal.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
+#include <wayland-server-core.h>
+#include <wlr/util/log.h>
+#include "rootston/desktop.h"
+
+static pid_t *im_pid;
+static const char *im_cmd;
+
+static void spawn_im_process() {
+	*im_pid = fork();
+	if (*im_pid == 0) {
+		sleep(1);
```

Comment:
sleep for what? If a sleep is needed for any reason it shoud likely be polling for something to be not racy

Reply:
Initially, I added the sleep there because when `spawn_im_process` is first called, the server is not completely initialized, causing an immediate crash. Later on, I realized this would be useful to rate-limit the spawns.

Reply:
-1

Rate-limiting shouldn't be done via `sleep`, and rootston shouldn't rate-limit. Child processes should only be spawned when the server is completely initialized.

Reply:
How to wait for full initialization?

Reply:
When I say rate-limit, I mean the situation where a faulty process will crash very soon after spawning and get restarted. I think it the infinite spawning loop should be prevented from excessively consuming resources, although not necessarily using sleep.

Reply:
You can either split your setup and call the spawn part directly before going into the main loop, or add an idle event at setup time to start the client from the mainloop.